File: ios/App/App/IsolatedModules/ProtocolWebView.swift
//
//  ProtocolWebView.swift
//  App
//
//  Created by Julia Samol on 06.10.22.
//

import Foundation
import Capacitor
import WebKit

class ProtocolWebView: NSObject, WKNavigationDelegate {
    private static let isolatedProtocolHTMLSource: String = "native/isolated_modules/isolated-protocol.html"
    
    private let webView: WKWebView
    private let jsAsyncResult: JSAsyncResult
    
    private var modules: [String: String] = [:]
    
    @MainActor
    init(assetsURL: URL) {
        let jsAsyncResult = JSAsyncResult()
        let userContentController = WKUserContentController()
        userContentController.add(jsAsyncResult)
        
        let webViewConfiguration = WKWebViewConfiguration()
        webViewConfiguration.userContentController = userContentController
        
        let webView = WKWebView(frame: .zero, configuration: webViewConfiguration)
        let _ = webView.loadFileURL(
            assetsURL.appendingPathComponent(Self.isolatedProtocolHTMLSource),
            allowingReadAccessTo: assetsURL
        )
        
        self.webView = webView
        self.jsAsyncResult = jsAsyncResult
        
        super.init()
        
        self.webView.navigationDelegate = self
    }
    
    func evaluateKeys(ofProtocol identifier: String, builtWithOptions options: JSObject?) async throws -> [String: Any] {
        try await evaluate(
            .keys,
            forProtocol: identifier,
            usingOptions: options
        )
    }
    
    func evaluateGetField(_ key: String, ofProtocol identifier: String, builtWithOptions options: JSObject?) async throws -> [String: Any] {
        try await evaluate(
            .getField(key),
            forProtocol: identifier,
            usingOptions: options
        )
    }
    
    func evaluateCallMethod(
        _ key: String,
        ofProtocol identifier: String,
        builtWithOptions options: JSObject?,
        withArgs args: JSArray?
    ) async throws -> [String: Any] {
        try await evaluate(
            .callMethod(key, args?.replaceNullWithUndefined()),
            forProtocol: identifier,
            usingOptions: options
        )
    }
    
    @MainActor
    private func evaluate(
        _ action: JSProtocolAction,
        forProtocol identifier: String,
        usingOptions options: JSObject?,
        injectScript script: String? = nil
    ) async throws -> [String: Any] {
        guard let module = loadModule(identifier) else {
            throw Error.moduleNotFound(identifier)
        }
        
        let resultID = await jsAsyncResult.createID()
        
        let script = """
            function postMessage(message) {
                window.webkit.messageHandlers.\(jsAsyncResult.id).postMessage({ ...message, id: "\(resultID)" });
            };
            
            execute(
                \(module),
                '\(identifier)',
                \(try options?.toJSONString() ?? JSUndefined.value.toJSONString()),
                \(try action.toJSONString()),
                function (result) {
                    postMessage({ result: JSON.parse(JSON.stringify(result)) });
                },
                function (error) {
                    postMessage({ error });
                }
            );
        """
        
        webView.evaluateJavaScript(script, completionHandler: nil)
        
        var result = try await jsAsyncResult.awaitResultWithID(resultID)
        result.replaceKey("result", with: action.resultField)
        
        return result
    }
    
    private func loadModule(_ identifier: String) -> String? {
        guard let moduleName = moduleName(fromIdentifier: identifier) else { return nil }
        if let module = modules[moduleName] { return module }
        
        guard let moduleSource = readModuleSource(moduleName) else { return nil }
        let jsModule = createJSModule(moduleName)
        modules[moduleName] = jsModule
        
        webView.evaluateJavaScript(moduleSource, completionHandler: nil)
        
        return jsModule
    }
    
    private func readModuleSource(_ name: String) -> String? {
        let publicURL = Bundle.main.url(forResource: "public", withExtension: nil)
        guard
            let sourcePath = publicURL?.appendingPathComponent("assets/libs/airgap-\(name).browserify.js"),
            let source = FileManager.default.contents(atPath: sourcePath.path)
        else { return nil }
        
        return .init(data: source, encoding: .utf8)
    }
    
    private func createJSModule(_ name: String) -> String {
        "airgapCoinLib\(name.capitalized)"
    }
    
    private func moduleName(fromIdentifier identifier: String) -> String? {
        switch identifier {
        case _ where identifier.starts(withAny: "ae"):
            return "aeternity"
        case _ where identifier.starts(withAny: "astar", "shiden"):
            return "astar"
        case _ where identifier.starts(withAny: "btc"):
            return "bitcoin"
        case _ where identifier.starts(withAny: "cosmos"):
            return "cosmos"
        case _ where identifier.starts(withAny: "eth"):
            return "ethereum"
        case _ where identifier.starts(withAny: "grs"):
            return "groestlcoin"
        case _ where identifier.starts(withAny: "moonbeam", "moonriver", "moonbase"):
            return "moonbeam"
        case _ where identifier.starts(withAny: "polkadot", "kusama"):
            return "polkadot"
        case _ where identifier.starts(withAny: "xtz"):
            return "tezos"
        default:
            return nil
        }
    }
    
    enum JSProtocolAction {
        private static let keysType: String = "keys"
        private static let getFieldType: String = "getField"
        private static let callMethodType: String = "callMethod"
        
        case keys
        case getField(String)
        case callMethod(String, JSArray?)
        
        var resultField: String {
            switch self {
            case .keys:
                return "keys"
            default:
                return "result"
            }
        }
        
        func toJSONString() throws -> String {
            switch self {
            case .keys:
                return """
                    {
                        "type": "\(Self.keysType)"
                    }
                """
            case let .getField(key):
                return """
                    {
                        "type": "\(Self.getFieldType)",
                        "field": "\(key)"
                    }
                """
            case let .callMethod(key, args):
                return """
                    {
                        "type": "\(Self.callMethodType)",
                        "method": "\(key)",
                        "args": \(try args?.toJSONString() ?? "[]")
                    }
                """
            }
        }
    }
    
    enum Error: Swift.Error {
        case moduleNotFound(String)
    }
}

private extension JSArray {
    func replaceNullWithUndefined() -> JSArray {
        map {
            if $0 is NSNull {
                return JSUndefined.value
            } else {
                return $0
            }
        }
    }
}

private extension Dictionary {
    mutating func replaceKey(_ oldKey: Key, with newKey: Key) {
        if let value = removeValue(forKey: oldKey) {
            self[newKey] = value
        }
    }
}

File: android/app/src/main/java/it/airgap/vault/plugin/isolatedmodules/ProtocolWebView.kt
package it.airgap.vault.plugin.isolatedmodules

import android.content.Context
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import it.airgap.vault.util.JSAsyncResult
import it.airgap.vault.util.JSUndefined
import it.airgap.vault.util.addJavascriptInterface
import it.airgap.vault.util.evaluateJavascript
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ProtocolWebView(context: Context) {
    private val jsAsyncResult: JSAsyncResult = JSAsyncResult()

    private val webViewLoadedDeferred: CompletableDeferred<Unit> = CompletableDeferred()
    private val webView: WebView = WebView(context).apply {
        visibility = View.GONE

        with(settings) {
            javaScriptEnabled = true
            allowFileAccess = false
        }

        addJavascriptInterface(jsAsyncResult)
        webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                webViewLoadedDeferred.complete(Unit)
            }
        }

        loadUrl("${ASSETS_PATH}/${ISOLATED_PROTOCOL_HTML_SOURCE}")
    }

    suspend fun evaluateKeys(identifier: String, options: JSObject?): JSObject =
        evaluate(identifier, options, JSProtocolAction.Keys)

    suspend fun evaluateGetField(identifier: String, options: JSObject?, key: String): JSObject =
        evaluate(identifier, options, JSProtocolAction.GetField(key))

    suspend fun evaluateCallMethod(identifier: String, options: JSObject?, key: String, args: JSArray?): JSObject =
        evaluate(identifier, options, JSProtocolAction.CallMethod(key, args?.replaceNullWithUndefined()))

    private suspend fun evaluate(
        identifier: String,
        options: JSObject?,
        action: JSProtocolAction,
    ): JSObject = withContext(Dispatchers.Main) {
        with(webView) {
            waitUntilLoaded()
            clear()

            val resultId = jsAsyncResult.createId()
            val script = """    
                execute(
                    '$identifier',
                    ${options.orUndefined()},
                    $action,
                    function (result) {
                        ${jsAsyncResult}.completeFromJS('$resultId', JSON.stringify({ ${action.resultField}: result }));
                    },
                    function (error) {
                        ${jsAsyncResult}.throwFromJS('$resultId', error);
                    }
                );
            """.trimIndent()

            evaluateJavascript(script)

            val result = jsAsyncResult.await(resultId).getOrThrow()
            JSObject(result).also { webView.clear() }
        }
    }

    private fun JSObject?.orUndefined(): Any = this ?: JSUndefined

    private fun JSArray.replaceNullWithUndefined(): JSArray =
        JSArray(toList<Any>().map { if (it == JSObject.NULL) JSUndefined else it })

    private suspend fun WebView.waitUntilLoaded() {
        webViewLoadedDeferred.await()
    }

    private fun WebView.clear() {
        clearCache(false)
        clearHistory()
        clearMatches()
        clearFormData()
        clearSslPreferences()
    }

    private sealed interface JSProtocolAction {
        val resultField: String
            get() = "result"

        object Keys : JSProtocolAction {
            private const val TYPE = "keys"
            override val resultField: String = "keys"

            override fun toString(): String = JSObject("""
                {
                    "type": "$TYPE"
                }
            """.trimIndent()).toString()
        }

        data class GetField(val key: String) : JSProtocolAction {
            override fun toString(): String = JSObject("""
                {
                    "type": "$TYPE",
                    "field": "$key"
                }
            """.trimIndent()).toString()

            companion object {
                private const val TYPE = "getField"
            }
        }

        data class CallMethod(val key: String, val args: JSArray?) : JSProtocolAction {
            override fun toString(): String {
                val args = args?.toString() ?: "[]"

                return JSObject("""
                    {
                        "type": "$TYPE",
                        "method": "$key",
                        "args": $args
                    }
                """.trimIndent()).toString()
            }

            companion object {
                private const val TYPE = "callMethod"
            }
        }
    }

    companion object {
        private const val ASSETS_PATH: String = "file:///android_asset"
        private const val ISOLATED_PROTOCOL_HTML_SOURCE: String = "public/assets/native/isolated_modules/isolated-protocol.html"
    }
}